### PR TITLE
Initial code to pull entries from another log.

### PR DESCRIPTION
**Pull request by pphaneuf**
This is a fairly complicated algorithm, so starting with a single peer, and I'll be building on that to add more features (multiple peers, continuous following, etc).

### CODE REVIEW

File: cpp/fetcher/fetcher.cc [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+using std::placeholders::_1;
+using std::unique_ptr;
+using std::vector;
+using util::Task;
+using util::TaskHold;
+
+namespace {
+
+
+const int kMaxFetches = 100;
```
Comment (AlCutter) [inline cpp/fetcher/fetcher.cc]: Make into flags?

File: cpp/fetcher/fetcher.cc [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    return;
+  }
+
+  int64_t index(start_);
+  int num_fetch(0);
+  for (Range* current = entries_.get(); current;
+       index += current->size_, current = current->next_.get()) {
+    lock_guard<mutex> lock(current->lock_);
+
+    // Coaslesce the next entries, if possible.
```
Comment (AlCutter) [inline cpp/fetcher/fetcher.cc]: "Coalesce with the next Range, if possible"?

File: cpp/fetcher/fetcher.cc [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+                << status;
+      range->state_ = Range::WANT;
+      return;
+    }
+
+    VLOG(1) << "received " << retval->size() << " entries at offset " << index;
+    int64_t processed(0);
+    for (const auto& entry : *retval) {
+      LoggedCertificate cert;
+      if (!cert.CopyFromClientLogEntry(entry)) {
```
Comment (AlCutter) [inline cpp/fetcher/fetcher.cc]: I think this probably merits at least a LOG(WARNING) or something, in all likelihood we're going to get wedged here on all subsequent runs (modulo a duff peer returning garbage etc.)

File: cpp/fetcher/remote_peer.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  // starts the asynchronous destruction).
+  RemotePeer(AsyncLogClient* client, LogVerifier* verifier, util::Task* task);
+  ~RemotePeer() override;
+
+  int64_t TreeSize() const override;
+
+ private:
+  struct Impl;
+
+  Impl* const impl_;
```
Comment (AlCutter) [inline cpp/fetcher/remote_peer.h]: Add a comment to say that this will be deleted by the Task

**Comment by pphaneuf** [comment]
I found a minor issue, when an issue is fetched twice, for some reason, I didn't check the error code for an existing entry, and just kept repeating... I'm running a full import of Aviator now to try to shake out issues, I'll address comments after that, in case anything else comes up.

File: cpp/fetcher/fetcher.cc (hunk 3 above)
Comment (pphaneuf) [inline cpp/fetcher/fetcher.cc]: This is where some of my current problems are, so I'll refine this a bit further, but I started out by adding log messages as suggested.

When this happens and we ```break;```, the logic below is what should avoid getting us wedged. On my TODO list for further pull requests is some handling for cases where we can't converge (say, a peer had a large tree size that we're trying to sync to, but it disappeared, so we can't possibly complete the operation).

**Comment by pphaneuf** [comment]
Fixed the spelling mistake there, and addressed other comments.

**Comment by AlCutter** [comment]
LGTM

**Pull request merged**